### PR TITLE
httputil: add NoNetwork(err) helper, spread test and use in serial acquire

### PR DESCRIPTION
**Pull request by mvo5**
This adds a new helper NoNetwork(err) and a spread test that
ensures that the errors behave sanely on all our tested systems.

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

File: tests/main/retry-network/task.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+summary: Test httputil/retry.go ShouldRetryError(err)/NoNetwork(err)
```
Comment (stolowski) [inline tests/main/retry-network/task.yaml]: That's a very neat idea \o/

File: tests/main/retry-network/task.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+restore: |
+    ip netns delete testns || true
+    umount /run/netns || true
+
+execute: |
+    echo "Add totally unconnected network namespace"
+    ip netns add testns
+    SNAPD_DEBUG=1 nsenter --net=/var/run/netns/testns ./detect-retry http://www.ubuntu.com > output.txt
+    #MATCH "ShouldRetryError: false" < output.txt
```
Comment (stolowski) [inline tests/main/retry-network/task.yaml]: Should it match ShouldRetryError: true? Or is it giving different result on different systems?

File: tests/main/retry-network/task.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    MATCH "NoNetwork: true" < output.txt
+
+    echo "Now without DNS resolver"
+    UBUNTU_IP="$(getent hosts www.ubuntu.com|awk '{print $1 }' | head -n1)"
+    SNAPD_DEBUG=1 nsenter --net=/var/run/netns/testns ./detect-retry "http://$UBUNTU_IP" > output.txt
+    #MATCH "ShouldRetryError: false" < output.txt
```
Comment (stolowski) [inline tests/main/retry-network/task.yaml]: Same question.

File: tests/main/retry-network/task.yaml (hunk 2 above)
Comment (mvo5) [reply tests/main/retry-network/task.yaml]: I added some more docstrings that hopefully answer this question. If not please let me know and I'm happy to elaborate further :)

File: httputil/retry.go [hunk 4]
```diff
@@ -98,7 +98,18 @@ func ShouldRetryAttempt(attempt *retry.Attempt, err error) bool {
 	return ShouldRetryError(err)
 }
 
+// ShouldRetryError returns true for transient network errors like when
+// the remote side returns a connection reset and it's sensible to rety
+// after a short time.
+//
+// XXX: Note that currently also NoNetwork(err) errors are repoarted
```
Comment (anonymouse64) [inline httputil/retry.go]:
```suggestion
// XXX: Note that currently also NoNetwork(err) errors are reported
```

File: overlord/devicestate/handlers_serial.go [hunk 5]
```diff
@@ -331,10 +331,19 @@ func prepareSerialRequest(t *state.Task, regCtx registrationContext, privKey ass
 	resp, err := client.Do(req)
 	if err != nil {
 		if !httputil.ShouldRetryError(err) {
-			// a non temporary net error, like a DNS no
-			// host, error out and do full retries
+			// a non temporary net error errors out and do full
```
Comment (anonymouse64) [inline overlord/devicestate/handlers_serial.go]:
```suggestion
			// a non temporary net error fully errors out and triggers a retry
```

File: tests/main/retry-network/task.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
+summary: Test httputil/retry.go ShouldRetryError(err)/NoNetwork(err)
+
+# ubuntu-core: no go-compiler
+# ubuntu-14.04: no nsenter
+systems: [-ubuntu-core-*, -ubuntu-14.04-*]
+
+environment:
+    # an empty $topsrcdir/tests/go.mod seems to break importing or building go
+    # packages referenced by their import paths while under the tests directory,
+    # need to disable go modules supportfor this test
```
Comment (anonymouse64) [inline tests/main/retry-network/task.yaml]:
```suggestion
    # need to disable go modules support for this test
```

Comment (anonymouse64) [inline tests/main/retry-network/task.yaml]: Other ways to handle this situation would be to just to add a minimal go.mod or use the go snap and install a Go version that doesn't care about go modules.

File: overlord/devicestate/handlers_serial.go [hunk 7]
```diff
@@ -331,10 +331,30 @@ func prepareSerialRequest(t *state.Task, regCtx registrationContext, privKey ass
...
 		}
+		if httputil.NoNetwork(err) {
+			// If there is no network there is no need to count
+			// this as a tentatives attempt. If we do it this
+			// way the risk is that we tried a bunch of times
+			// with no network and if we hit the server for real
+			// and it replies with something we need to retry
+			// we will not because nTentatives is way over the
+			// limit.
+			nTentatives--
```
Comment (bboozzoo) [inline overlord/devicestate/handlers_serial.go]: Shouldn't we reset pre-poll-tenantives to 0?

File: httputil/retry.go [hunk 8]
```diff
@@ -98,7 +98,18 @@ func ShouldRetryAttempt(attempt *retry.Attempt, err error) bool {
...
+// the remote side returns a connection reset and it's sensible to rety
+// after a short time.
+//
+// XXX: Note that currently also NoNetwork(err) errors are reported
+// with true here.
 func ShouldRetryError(err error) bool {
+	if err == nil {
+		return false
+	}
+	logger.Debugf("ShouldRetryError: %v %T", err, err)
```
Comment (pedronis) [inline httputil/retry.go]: for debug shouldn't we maybe try to log the result as well?

File: httputil/retry.go [hunk 9]
```diff
@@ -174,7 +185,18 @@ func ShouldRetryError(err error) bool {
 	return false
 }
 
+// NoNetwork returns true if the error indicates that there is no network
+// connection available, i.e. network unreachable or down or DNS unavailable.
+func NoNetwork(err error) bool {
+	logger.Debugf("NoNetwork: %v %T", err, err)
```
Comment (pedronis) [inline httputil/retry.go]: same

File: overlord/devicestate/handlers_serial.go (hunk 7 above)
Comment (pedronis) [reply overlord/devicestate/handlers_serial.go]: yes, that seems easier to reason about

**Pull request merged**